cli/data.go

```diff
@@ -595,6 +595,131 @@ func (c *viamClient) dataAddToDataset(datasetID, orgID, locationID string, fileI
+) error {
+	ids := make(chan *datapb.BinaryID, parallelActions)
+	// Give channel buffer of 1+parallelActions because that is the number of goroutines that may be passing an
+	// error into this channel (1 get ids routine + parallelActions download routines).
+	errs := make(chan error, 1+parallelActions)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	var wg sync.WaitGroup
+
+	// In one routine, get all IDs matching the filter and pass them into ids.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		// If limit is too high the request can time out, so limit each call to a maximum value of 100.
+		var limit uint
+		if parallelActions > maxLimit {
+			limit = maxLimit
+		} else {
+			limit = parallelActions
+		}
+		if err := getMatchingBinaryIDs(ctx, dataClient, filter, ids, limit); err != nil {
+			errs <- err
+			cancel()
+		}
+	}()
+
+	// In parallel, read from ids and perform the action on the binary data for each id in batches of parallelActions.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		var nextID *datapb.BinaryID
+		var done bool
+		var numFilesAdded atomic.Int32
+		var downloadWG sync.WaitGroup
+		for {
+			for i := uint(0); i < parallelActions; i++ {
+				if err := ctx.Err(); err != nil {
+					errs <- err
+					cancel()
+					done = true
+					break
+				}
+
+				nextID = <-ids
+
+				// If nextID is nil, the channel has been closed and there are no more IDs to be read.
+				if nextID == nil {
+					done = true
+					break
+				}
+
+				downloadWG.Add(1)
+				go func(id *datapb.BinaryID) {
+					defer downloadWG.Done()
+					err := actionOnBinaryData(id)
+					if err != nil {
+						errs <- err
+						cancel()
+						done = true
+					}
+					numFilesAdded.Add(1)
+					if numFilesAdded.Load()%logEveryN == 0 {
+						printf(writer, "Added %d files", numFilesAdded.Load())
+					}
+				}(nextID)
+			}
+			downloadWG.Wait()
+			if done {
+				break
+			}
+		}
+		if numFilesAdded.Load()%logEveryN != 0 {
+			printf(writer, "Added %d files", numFilesAdded.Load())
+		}
+	}()
+	wg.Wait()
+	close(errs)
+
+	if err := <-errs; err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func addDataToDatasetByIDWrapper(ctx context.Context, dataClient datapb.DataServiceClient,
+	datasetID string, id *datapb.BinaryID,
+) error {
+	_, err := dataClient.AddBinaryDataToDatasetByIDs(ctx,
+		&datapb.AddBinaryDataToDatasetByIDsRequest{DatasetId: datasetID, BinaryIds: []*datapb.BinaryID{id}})
+	return err
+}
+
+// dataAddToDatasetByFilter adds data, with the specified filter to the dataset corresponding to the dataset ID.
+func (c *viamClient) dataAddToDatasetByFilter(filter *datapb.Filter, datasetID string) error {
+	if err := c.ensureLoggedIn(); err != nil {
+		return err
+	}
+	parallelActions := uint(100)
+
+	return performActionOnBinaryDataFromFilter(
+		func(id *datapb.BinaryID) error {
+			return addDataToDatasetByIDWrapper(c.c.Context, c.dataClient, datasetID, id)
```

Comment:
[nit/optional] for readability, I might just add the code inline here instead of wrapping

```
_, err := c.dataClient.AddBinaryDataToDatasetByIDs(c.c.Context,
  &datapb.AddBinaryDataToDatasetByIDsRequest{DatasetId: datasetID, BinaryIds: []*datapb.BinaryID{id}})
return err
```

Reply:
Updated